src/runtime_src/core/common/info_memory.cpp

```diff
@@ -411,4 +413,190 @@ xclbin_info(const xrt_core::device * device)
+    xrt_core::scope_guard<std::function<void()>> g(std::bind(at_exit, dev, uuid));
+
+    dev->update_scheduler_status();
+  }
+  catch (const std::exception&) {
+    // xclbin_lock failed, safe to ignore
+  }
+}
+
+std::vector<ps_kernel_data> 
+get_ps_kernels(const xrt_core::device *device)
+{
+  std::vector<ps_kernel_data> ps_kernels;
+  try {
+    std::vector<char> buf = xrt_core::device_query<xq::ps_kernel>(device);
+    if (buf.empty())
+      return ps_kernels;
+    const ps_kernel_node *map = reinterpret_cast<ps_kernel_node *>(buf.data());
+    if(map->pkn_count < 0)
+      throw xrt_core::error("'ps_kernel' invalid. Has the PS kernel been loaded? See 'xbutil program'.");
+
+    for (unsigned int i = 0; i < map->pkn_count; i++)
+      ps_kernels.emplace_back(map->pkn_data[i]);
+  }
+  catch (const xq::no_such_key&) {
+    // Ignoring if not available: Edge Case
+  }
+
+  return ps_kernels;
+}
+
+ptree_type
+populate_cus(const xrt_core::device *device)
+{
+  scheduler_update_stat(device);
+
+  ptree_type pt;
+  using cu_data_type = xq::kds_cu_info::data_type;
+  using scu_data_type = xq::kds_scu_info::data_type;
+  std::vector<cu_data_type> cu_stats;
+  std::vector<scu_data_type> scu_stats;
+  ptree_type ptree;
+  try {
+    std::string uuid = xrt_core::device_query<xq::xclbin_uuid>(device);
+    boost::algorithm::to_upper(uuid);
+    ptree.put("xclbin_uuid", uuid);
+  } catch (xq::exception&) {  }
+
+  try {
+    cu_stats  = xrt_core::device_query<xq::kds_cu_info>(device);
+    scu_stats = xrt_core::device_query<xq::kds_scu_info>(device);
+  }
+  catch (const xq::no_such_key&) {
+    // Ignoring if not available: Edge Case
+  }
+  catch (const std::exception& ex) {
+    ptree.put("error_msg", ex.what());
+    return ptree;
+  }
+
+  for (auto& stat : cu_stats) {
+    ptree_type pt_cu;
+    pt_cu.put( "name",           stat.name);
+    pt_cu.put( "base_address",   boost::str(boost::format("0x%x") % stat.base_addr));
+    pt_cu.put( "usage",          stat.usages);
+    pt_cu.put( "type", enum_to_str(cu_type::pl));
+    pt_cu.add_child( std::string("status"),	get_cu_status(stat.status));
+    pt.push_back(std::make_pair("", pt_cu));
+  }
+
+  std::vector<ps_kernel_data> ps_kernels;
+  try {
+    ps_kernels = get_ps_kernels(device);
+  } catch(const xrt_core::error& ex) {
+    std::cout << ex.what() <<std::endl;
+    return ptree;
+  }
+
+  uint32_t psk_inst = 0;
+  uint32_t num_scu = 0;
+  ptree_type pscu_list;
+  for (auto& stat : scu_stats) {
+    ptree_type pt_cu;
+    std::string scu_name = "Illegal";
+    if (psk_inst >= ps_kernels.size()) {
+      scu_name = stat.name;
+      //This means something is wrong
+      //scu_name e.g. kernel_vcu_encoder:scu_34
+    } else {
+      scu_name = ps_kernels.at(psk_inst).pkd_sym_name;
+      scu_name.append("_");
+      scu_name.append(std::to_string(num_scu));
+      //scu_name e.g. kernel_vcu_encoder_2
```

Comment:
Hmm...what is the purpose of this comment?

Reply:
This means what a valid scu name would look like and what are we parsing in the else statement